pkg/user/tx_client.go

```diff
@@ -166,83 +170,124 @@ func SetupTxClient(
+	blobSizes := make([]uint32, len(blobs))
+	for i, blob := range blobs {
+		blobSizes[i] = uint32(len(blob.Data))
+	}
+
+	gasLimit := uint64(float64(types.DefaultEstimateGas(blobSizes)) * client.gasMultiplier)
+	fee := uint64(math.Ceil(appconsts.DefaultMinGasPrice * float64(gasLimit)))
+	// prepend calculated params, so it can be overwritten in case the user has specified it.
+	opts = append([]TxOption{SetGasLimit(gasLimit), SetFee(fee)}, opts...)
+
+	txBytes, _, err := client.signer.CreatePayForBlobs(account, blobs, opts...)
 	if err != nil {
 		return nil, err
 	}
 
-	return s.broadcastTx(ctx, txBytes, account)
+	return client.broadcastTx(ctx, txBytes, account)
 }
 
 // SubmitTx forms a transaction from the provided messages, signs it, and submits it to the chain. TxOptions
 // may be provided to set the fee and gas limit.
-func (s *TxClient) SubmitTx(ctx context.Context, msgs []sdktypes.Msg, opts ...TxOption) (*sdktypes.TxResponse, error) {
-	resp, err := s.BroadcastTx(ctx, msgs, opts...)
+func (client *TxClient) SubmitTx(ctx context.Context, msgs []sdktypes.Msg, opts ...TxOption) (*sdktypes.TxResponse, error) {
+	resp, err := client.BroadcastTx(ctx, msgs, opts...)
 	if err != nil {
 		return resp, err
 	}
 
-	return s.ConfirmTx(ctx, resp.TxHash)
+	return client.ConfirmTx(ctx, resp.TxHash)
 }
 
-func (s *TxClient) BroadcastTx(ctx context.Context, msgs []sdktypes.Msg, opts ...TxOption) (*sdktypes.TxResponse, error) {
-	s.mtx.Lock()
-	defer s.mtx.Unlock()
-	account, err := s.getAccountNameFromMsgs(msgs)
+func (client *TxClient) BroadcastTx(ctx context.Context, msgs []sdktypes.Msg, opts ...TxOption) (*sdktypes.TxResponse, error) {
+	client.mtx.Lock()
+	defer client.mtx.Unlock()
+	account, err := client.getAccountNameFromMsgs(msgs)
 	if err != nil {
 		return nil, err
 	}
 
-	if err := s.checkAccountLoaded(ctx, account); err != nil {
+	if err := client.checkAccountLoaded(ctx, account); err != nil {
 		return nil, err
 	}
 
-	tx, account, _, err := s.signer.SignTx(msgs, opts...)
+	txBuilder, err := client.signer.txBuilder(msgs, opts...)
 	if err != nil {
 		return nil, err
 	}
 
-	txBytes, err := s.signer.EncodeTx(tx)
+	estimated := false
+	for _, coin := range txBuilder.GetTx().GetFee() {
+		if coin.Denom == appconsts.BondDenom {
+			estimated = true
+			break
+		}
+	}
+
+	gasLimit := txBuilder.GetTx().GetGas()
+	if float32(gasLimit) < appconsts.DefaultMinGasPrice {
+		if !estimated {
+			// add at least 1utia as fee to builder as it affects gas calculation.
+			txBuilder.SetFeeAmount(sdktypes.NewCoins(sdktypes.NewCoin(appconsts.BondDenom, sdktypes.NewInt(1))))
+		}
+		gasLimit, err = client.estimateGas(ctx, txBuilder)
+		if err != nil {
+			return nil, err
+		}
+		txBuilder.SetGasLimit(gasLimit)
+	}
+
+	if !estimated {
+		fee := int64(math.Ceil(appconsts.DefaultMinGasPrice * float64(gasLimit)))
+		txBuilder.SetFeeAmount(sdktypes.NewCoins(sdktypes.NewCoin(appconsts.BondDenom, sdktypes.NewInt(fee))))
```

Comment:
I think there's another option like `SetFee` where you don't need to specify the `BondDenom`

Reply:
The option you've mentioned is a TxOption, that invokes `builder.SetFeeAmount `with utia as denomination under the hood.

Reply:
Ah I get it.